### PR TITLE
CalDAV: support PROPPATCH for default-scheduling-calendar property

### PR DESCRIPTION
**Pull request by rsto**
There's Cassandane tests at https://github.com/cyrusimap/cassandane/tree/caldav_scheddefault_move

### CODE REVIEW

File: imap/http_dav.c [hunk 1]
```diff
@@ -4974,33 +4976,52 @@ int meth_delete(struct transaction_t *txn, void *params)
...
+                txn->error.precond = DAV_NEED_LOCK_TOKEN;
+                txn->error.resource = txn->req_tgt.path;
+                GCC_FALLTHROUGH
+            default:
+                /* We failed a precondition - don't perform the request */
+                r = ret ? ret : HTTP_SERVER_ERROR;
+                goto done;
+        }
+
         if (dparams->delete) {
```
Comment (ksmurchison) [inline imap/http_dav.c:5005]: As per my comment above, we would NOT make the change above.

File: imap/http_dav.c [hunk 2]
```diff
@@ -4974,33 +4976,52 @@ int meth_delete(struct transaction_t *txn, void *params)
...
-            /* Open mailbox for reading */
-            r = mailbox_open_irl(txn->req_tgt.mbentry->name, &mailbox);
-            if (r) {
-                syslog(LOG_ERR, "http_mailbox_open(%s) failed: %s",
-                       txn->req_tgt.mbentry->name, error_message(r));
-                txn->error.desc = error_message(r);
-                return HTTP_SERVER_ERROR;
-            }
-
             /* Open the DAV DB corresponding to the mailbox */
```
Comment (ksmurchison) [inline imap/http_dav.c:5009]: As per my comment above, we would NOT make the change above and we would do the following before opening DAV DB:

ret = dparams->delete(txn, mailbox, NULL, NULL);
if (ret) goto done;

File: imap/http_dav.c [hunk 3]
```diff
@@ -4974,33 +4976,52 @@ int meth_delete(struct transaction_t *txn, void *params)
...
-                return HTTP_SERVER_ERROR;
+                ret = HTTP_SERVER_ERROR;
+                goto done;
             }
         }
 
+        /* we need the mailbox closed before we delete it */
+        mailbox_close(&mailbox);
+
         mbname_t *mbname = mbname_from_intname(txn->req_tgt.mbentry->name);
```
Comment (ksmurchison) [inline imap/http_dav.c:5025]: As per my comment above, we would NOT make these changes.

File: imap/http_caldav.c [hunk 4]
```diff
@@ -1190,6 +1194,126 @@ static modseq_t caldav_get_modseq(struct mailbox *mailbox,
...
+        }
+    }
+
+    if (mbname) {
+        char *calhomename = caldav_mboxname(httpd_userid, NULL);
+        struct mailbox *calhome = NULL;
+        struct mailbox *mailbox = NULL;
+        int r = mailbox_open_iwl(calhomename, &calhome);
+        if (!r) r = mailbox_open_iwl(mbname_intname(mbname), &mailbox);
+        if (!r) {
```
Comment (ksmurchison) [inline imap/http_caldav.c:1286]: Why do we need to open the prospective default calendar if we're setting the annotation on cal-home-set?

Comment (rsto) [reply imap/http_caldav.c:1286]: We don't want anyone deleting it before we are finished with the transaction.

Comment (ksmurchison) [reply imap/http_caldav.c:1286]: Yes, make sense.  I didn't fully form my previous thought.  Would a namelock be cheaper than opening and closing the mailbox?

File: imap/http_caldav.c [hunk 5]
```diff
@@ -1200,7 +1324,20 @@ static int caldav_check_precond(struct transaction_t *txn,
...
+        char *defaultname = caldav_scheddefault(httpd_userid);
+        char *defaultmboxname = caldav_mboxname(httpd_userid, defaultname);
+        if (!strcmp(mailbox->name, defaultmboxname)) {
+            precond = HTTP_FORBIDDEN;
+            txn->error.precond = CALDAV_DEFAULT_NEEDED;
+        }
+        free(defaultmboxname);
+        free(defaultname);
+        if (precond) return precond;
+    }
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_caldav.c:1341]: ksmurchison and discussed this off Github and agreed we keep it in check_precond for now.

File: imap/http_caldav.c (hunk 4 above)
Comment (rsto) [reply imap/http_caldav.c:1286]: We need to lock the mailbox, the namelock is not enough.

File: imap/http_caldav.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_caldav.c:1341]: ksmurchison had a discussion off Github and decided to keep this in check_precond for now. We'll refactor that part of the DAV code later.

File: imap/http_caldav.c (hunk 4 above)
Comment (rsto) [reply imap/http_caldav.c:1286]: We need to lock the mailbox, a namelock is not enough.

**Pull request merged**